src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,5392 @@
+
+static int
+new_fxstat(int vers, int fd, struct stat *buf)
+{
+	int rc;
+
+	if (!hook_enabled || fd < FD_FILE_BASE)
+		return next_fxstat(vers, fd, buf);
+
+	if (fd < FD_DIR_BASE) {
+		rc          = dfs_ostat(file_list[fd - FD_FILE_BASE]->dfs_mt->dfs,
+					file_list[fd - FD_FILE_BASE]->file, buf);
+		buf->st_ino = file_list[fd - FD_FILE_BASE]->st_ino;
+	} else {
+		rc          = dfs_ostat(dir_list[fd - FD_DIR_BASE]->dfs_mt->dfs,
+					dir_list[fd - FD_DIR_BASE]->dir, buf);
+		buf->st_ino = dir_list[fd - FD_DIR_BASE]->st_ino;
+	}
+
+	if (rc) {
+		if (bLog)
+			D_ERROR("Failed to call dfs_ostat. %s\n", strerror(rc));
+		errno = rc;
+		rc    = -1;
+	}
+
+	if (bLog)
+		atomic_fetch_add_relaxed(&num_stat, 1);
+
+	return 0;
+}
+
+static int
+new_xstat(int ver, const char *path, struct stat *stat_buf)
+{
+	int              is_target_path, rc;
+	dfs_obj_t       *parent;
+	dfs_obj_t       *obj;
+	struct dfs_mt   *dfs_mt;
+	mode_t           mode;
+	char             item_name[DFS_MAX_NAME];
+	char             *parent_dir = NULL;
+	char             *full_path  = NULL;
+
+	if (!hook_enabled)
+		return next_xstat(ver, path, stat_buf);
+
+	rc = query_path(path, &is_target_path, &parent, item_name, &parent_dir,
+			&full_path, &dfs_mt);
+	if (rc)
+		D_GOTO(out_err, rc = errno);
+	if (!is_target_path)
+		goto out_org;
+	if (bLog)
+		atomic_fetch_add_relaxed(&num_stat, 1);
+
+	if (!parent && (strncmp(item_name, "/", 2) == 0)) {
+		rc = dfs_lookup(dfs_mt->dfs, "/", O_RDONLY, &obj, &mode, stat_buf);
+	} else {
+		rc = dfs_lookup_rel(dfs_mt->dfs, parent, item_name, O_RDONLY, &obj, &mode,
+				    stat_buf);
+	}
+	stat_buf->st_mode = mode;
+	if (rc)
+		D_GOTO(out_err, rc);
+
+	stat_buf->st_ino = FAKE_ST_INO(full_path);
+	dfs_release(obj);
+	if (S_ISDIR(mode))
+		/* st_nlink should be 2 or larger number for dir */
+		stat_buf->st_nlink = 2;
+	FREE(parent_dir);
+
+	return 0;
+
+out_org:
+	FREE(parent_dir);
+	return next_xstat(ver, path, stat_buf);
+
+out_err:
+	FREE(parent_dir);
+	errno = rc;
+	return (-1);
+}
+
+static int
+new_lxstat(int ver, const char *path, struct stat *stat_buf)
+{
+	int              is_target_path, rc;
+	dfs_obj_t       *parent;
+	struct dfs_mt   *dfs_mt;
+	char             item_name[DFS_MAX_NAME];
+	char             *parent_dir = NULL;
+	char             *full_path  = NULL;
+
+	if (!hook_enabled)
+		return libc_lxstat(ver, path, stat_buf);
+
+	rc = query_path(path, &is_target_path, &parent, item_name, &parent_dir,
+			&full_path, &dfs_mt);
+	if (rc)
+		D_GOTO(out_err, rc = errno);
+	if (!is_target_path)
+		goto out_org;
+	if (bLog)
+		atomic_fetch_add_relaxed(&num_stat, 1);
+
+	if (!parent && (strncmp(item_name, "/", 2) == 0))
+		rc = dfs_stat(dfs_mt->dfs, NULL, NULL, stat_buf);
+	else
+		rc = dfs_stat(dfs_mt->dfs, parent, item_name, stat_buf);
+	if (rc)
+		goto out_err;
+	stat_buf->st_ino = FAKE_ST_INO(full_path);
+	if (S_ISDIR(stat_buf->st_mode))
+		/* st_nlink should be 2 or larger number for dir in rm */
```

Comment:
I wasn't aware of nlink wanting to be 2 or larger, what's the relationship with rm?

Reply:
In my impression, I found a piece of code in "rm" also checks whether st_nlink is larger than 1 to tell whether a path is a directory. I could not find the exact place today. I will remove this code now. Thank you!